### PR TITLE
#630 ReaderOf test class is incomplete

### PR DESCRIPTION
**Pull request by proshin-roman**
Pull request for the issue #630. It increases test coverage of ReaderOf class up to 100%. All ctors are covered by successful tests now. No other changes.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#676](https://github.com/yegor256/cactoos/issues/676) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#676](https://github.com/yegor256/cactoos/issues/676) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/491053620). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/test/java/org/cactoos/io/ReaderOfTest.java [hunk 1]
```diff
@@ -40,18 +41,317 @@
...
 public final class ReaderOfTest {
 
     @Test
-    public void readsSimpleFileContent() throws IOException {
-        final Path temp = Files.createTempFile("cactoos-1", "txt-1");
+    public void readsCharVarArg() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't read content",
+            new InputOf(new ReaderOf('a', 'b', 'c')),
+            new InputHasContent("abc")
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/ReaderOfTest.java]: @proshin-roman Is it not possible to use Hamcrest matchers here? It seems strange to me, to use matchers from cactoos itself -- what if the matcher will have a bug in the future? Tests will break for that reason and it will be misleading. 

I would use trusted Hamcrest matchers, even if you have to call ``asString()`` or something, on ``InputOf``

Furthermore, it would be nice if we could test ``ReaderOf`` without having to wrap it inside ``InputOf`` -- what if the test fails because there is a bug in ``InputOf`` and ``ReaderOf`` is actually fine?

File: src/test/java/org/cactoos/io/ReaderOfTest.java [hunk 2]
```diff
@@ -40,18 +41,317 @@
...
+            "Can't read content",
+            new InputOf(
+                new ReaderOf(
+                    new InputStreamOf("Hello, товарищ!"),
+                    StandardCharsets.UTF_8.newDecoder()
+                )
+            ),
+            new InputHasContent("Hello, товарищ!")
+        );
+    }
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/ReaderOfTest.java]: @proshin-roman Can you add a few tests with empty and null content? What happens if ReaderOf tries to read from an empty/null source?

File: src/test/java/org/cactoos/io/ReaderOfTest.java [hunk 3]
```diff
@@ -40,18 +41,317 @@
...
 }
```
Comment (amihaiemil) [inline src/test/java/org/cactoos/io/ReaderOfTest.java]: @proshin-roman Can you also add a test for german characters? It is also quite a spoken language, so we should be able to read german diacritics. The german diacritics are: ``äÄ üÜ öÖ and ß``

File: src/test/java/org/cactoos/io/ReaderOfTest.java [hunk 4]
```diff
@@ -40,18 +41,317 @@
  * @checkstyle JavadocMethodCheck (500 lines)
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
+@SuppressWarnings({"PMD.TooManyMethods", "PMD.AvoidDuplicateLiterals"})
```
[the remark this answers is not in the document]
Comment (proshin-roman) [reply src/test/java/org/cactoos/io/ReaderOfTest.java]: I've extracted test content into a final static (!) field and got rid of other string constants.

File: src/test/java/org/cactoos/io/ReaderOfTest.java (hunk 1 above)
Comment (proshin-roman) [reply src/test/java/org/cactoos/io/ReaderOfTest.java]: I've tried to do that using one private method that does the assertion.

File: src/test/java/org/cactoos/io/ReaderOfTest.java (hunk 2 above)
Comment (proshin-roman) [reply src/test/java/org/cactoos/io/ReaderOfTest.java]: I've added tests for null and empty string.

File: src/test/java/org/cactoos/io/ReaderOfTest.java (hunk 3 above)
Comment (proshin-roman) [reply src/test/java/org/cactoos/io/ReaderOfTest.java]: That was done too.

**Comment by amihaiemil** [comment]
@proshin-roman Looks ok to me, but I believe the architect will complain about that ``final static CONTENT``, since it's a shared resource between tests. [Here](http://www.yegor256.com/2016/05/03/test-methods-must-share-nothing.html) is the article about it. 

Let's see what he says though :)

**Comment by amihaiemil** [comment]
@rultor Good to merge

**Comment by amihaiemil** [comment]
@proshin-roman And please, edit all your comments (including review comments) and address them to me (they should start with ``@amihaiemil``) -- the QA will complain about this stuff :D

**Comment by proshin-roman** [comment]
@amihaiemil Exactly! I know about "no static" rule, but I believe it's ok for this test as it's really same content for all the test cases. Theoretically, we can avoid that by using something like "<current string> + <index>" so that we have technically different strings, but it doesn't change the thing.

**Comment by amihaiemil** [comment]
@yegor256 ping

File: src/test/java/org/cactoos/io/ReaderOfTest.java [hunk 5]
```diff
@@ -40,18 +45,307 @@
...
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
+@SuppressWarnings("PMD.TooManyMethods")
 public final class ReaderOfTest {
 
+    /**
+     * Test content for all the tests in this class.
+     */
+    private static final String CONTENT =
+        "Hello, товарищ äÄ üÜ öÖ and ß";
```
Comment (yegor256) [inline src/test/java/org/cactoos/io/ReaderOfTest.java]: @proshin-roman it's a bad idea, see http://www.yegor256.com/2016/05/03/test-methods-must-share-nothing.html

**Comment by yegor256** [comment]
@proshin-roman see above

**Comment by proshin-roman** [comment]
@yegor256 @amihaiemil I've updated pull request: no static fields now. Please review it again.

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 we can merge this now, I think

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by ypshenychka** [comment]
@amihaiemil quality good

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
`QA` review completed: +8 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+694](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
Order was finished, quality was "good": +20 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+980](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/yegor256/cactoos/issues/676#issuecomment-368480895))

@ypshenychka You can't do that, unless you have one of these roles: `ARC`, `PO`. Your current roles are: `QA`.

**Comment by 0crat** [comment]
This pull request [#676](https://github.com/yegor256/cactoos/issues/676) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/591809452). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by neonailol** [comment]
@0crat status

**Comment by 0crat** [comment]
Are you speaking to me or about me [here](https://github.com/yegor256/cactoos/issues/676#issuecomment-369838351). You must always start your message with my name if you want to address it to me, see [§1](http://www.zerocracy.com/policy.html#1).

**Comment by neonailol** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/676#issuecomment-370381498))

@neonailol This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#676](https://github.com/yegor256/cactoos/issues/676) is [in scope](http://www.zerocracy.com/policy.html#14) for 15days
  * The role is `REV`
  * The job is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol) for [5days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @neonailol[/z](https://www.0crat.com/u/neonailol) will get 25% of their hourly rate for this job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @proshin-roman[/z](https://www.0crat.com/u/proshin-roman): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%23676%27%7D) (restricted area)

**Comment by neonailol** [comment]
@yegor256 this job was assigned to me by mistake

**Comment by yegor256** [comment]
@0crat refuse

**Comment by yegor256** [comment]
@0crat out

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/676#issuecomment-370389066))

@yegor256 The user @neonailol[/z](https://www.0crat.com/u/neonailol) resigned from [#676](https://github.com/yegor256/cactoos/issues/676), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/yegor256/cactoos/issues/676#issuecomment-370389086))

@yegor256 The job [#676](https://github.com/yegor256/cactoos/issues/676) is now out of scope
